File: 0101-0200/0111.Minimum_Depth_of_Binary_Tree.java
//0ms

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int level = 0, min = Integer.MAX_VALUE;
    public void getMin(TreeNode root) {
        ++level;
        if (level == min) {
            --level;
            return;
        }
        if (root.left == null && root.right == null) {
            min = Math.min(level, min);
            --level;
            return;
        }
        if ((root.left != null) && (root.right == null || root.right.right != null || root.right.left != null)) getMin(root.left);
        if (root.right != null) getMin(root.right);
        --level;
    }
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        getMin(root);
        return min;
    }
}

File: 0101-0200/0111.Minimum_Depth_of_Binary_Tree.rs
//0ms

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn min_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(node: &Option<Rc<RefCell<TreeNode>>>, dsf: i32) -> i32 {
            if let Some(node) = node {
                let left = dfs(&node.borrow_mut().left.take(), dsf + 1);
                let right = dfs(&node.borrow_mut().right.take(), dsf + 1);
                if left == dsf + 1 {
                    return right;
                } else if right == dsf + 1 {
                    return left;
                }
                return std::cmp::min(left, right);
            }
            dsf
        }
        dfs(&root, 0)
    }
}
